src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+
+        public ReadOnlySpan<byte> Bytes => _buffer;
+
+        public override string ToString()
+        {
+            var status = Encodings.Utf8.ToUtf16Length(this.Bytes, out int needed);
+            if (status != Buffers.OperationStatus.Done)
+                return string.Empty;
+
+            // UTF-16 is 2 bytes per char
+            var chars = new char[needed >> 1];
+            var utf16 = new Span<char>(chars).AsBytes();
+            status = Encodings.Utf8.ToUtf16(this.Bytes, utf16, out int consumed, out int written);
+            if (status != Buffers.OperationStatus.Done)
+                return string.Empty;
+
+            return new string(chars);
+        }
+
+        public bool ReferenceEquals(Utf8Span other)
+        {
+            return _buffer == other._buffer;
+        }
+
+        public bool Equals(Utf8Span other)
+        {
+            return _buffer.SequenceEqual(other._buffer);
+        }
+
+        public bool Equals(string other)
+        {
+            CodePointEnumerator thisEnumerator = GetCodePointEnumerator();
+            Utf16LittleEndianCodePointEnumerator otherEnumerator = new Utf16LittleEndianCodePointEnumerator(other);
+
+            while (true)
+            {
+                bool hasNext = thisEnumerator.MoveNext();
+                if (hasNext != otherEnumerator.MoveNext())
+                {
+                    return false;
+                }
+
+                if (!hasNext)
+                {
+                    return true;
+                }
+
+                if (thisEnumerator.Current != otherEnumerator.Current)
+                {
+                    return false;
+                }
+            }
+        }
+
+        public static bool operator ==(Utf8Span left, Utf8Span right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Utf8Span left, Utf8Span right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static bool operator ==(Utf8Span left, string right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Utf8Span left, string right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static bool operator ==(string left, Utf8Span right)
+        {
+            return right.Equals(left);
+        }
+
+        public static bool operator !=(string left, Utf8Span right)
+        {
+            return !right.Equals(left);
+        }
+
+        public int CompareTo(Utf8Span other)
+        {
+            throw new NotImplementedException();
+        }
+
+        public int CompareTo(string other)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="index">Index in UTF-8 code units (bytes)</param>
+        /// <returns>Length in UTF-8 code units (bytes)</returns>
+        public Utf8Span Substring(int index)
```

Comment:
I think it's ok to merge this for now, but I think we need to decide what to do with all the index-based APIs on Utf8 strings. I am not sure if indexing based on code units is useful. Maybe only deal with code points in Utf8Strings.

Reply:
I think it's ok to leave it for now and do a holistic cleanup of the index based APIs.